lib/ansible/modules/packaging/os/package_facts.py

```diff
@@ -140,94 +145,229 @@
+from ansible.module_utils._text import to_native, to_text
+
+
+class PkgMgr:
+
+    def __init__(self, module):
+
+        self.m = module
+
+    def test(self):
+        ''' This method is supposed to return True/False if the package manager is installed/usable by this module '''
+        raise NotImplementedError
+
+    def list_installed(self):
+        ''' This method should return a list of installed packages, each list item will be passed to get_package_details '''
+        raise NotImplementedError
+
+    def get_package_details(self, package):
+        ''' This takes a 'package' item and returns a dictionary with the package information, name and version are minimal requirements '''
+        raise NotImplementedError
+
+    def get_packages(self):
+        ''' Take all of the above and return a dictionary of lists of dictionaries (package = list of installed versions) '''
+
+        installed_packages = {}
+        for package in self.list_installed():
+            package_details = self.get_package_details(package)
+            package_details['source'] = self.__class__.__name__.lower()
+            name = package_details['name']
+            if name not in installed_packages:
+                installed_packages[name] = [package_details]
+            else:
+                installed_packages[name].append(package_details)
+        return installed_packages
+
+
+class LibMgr(PkgMgr):
+
+    LIB = None
+
+    def __init__(self, module):
+
+        super(LibMgr, self).__init__(module)
+        self.lib = None
+
+    def test(self):
+        found = False
+        try:
+            self.lib = __import__(self.LIB)
+            found = True
+        except ImportError:
+            pass
+        return found
+
+
+class CLIMgr(PkgMgr):
+
+    CLI = None
+
+    def __init__(self, module):
+
+        super(CLIMgr, self).__init__(module)
+        self.cli = None
+
+    def test(self):
+        self.cli = self.m.get_bin_path(self.CLI, False)
+        return bool(self.cli)
+
+
+class APT(LibMgr):
+
+    LIB = 'apt'
+
+    def test(self):
+        works = super(APT, self).test()
+        if works:
+            self._cache = self.lib.Cache()
+        return works
+
+    def list_installed(self):
+        return [pk for pk in self._cache.keys() if self._cache[pk].is_installed]
+
+    def get_package_details(self, package):
+        ac_pkg = self._cache[package].installed
+        return dict(name=package, version=ac_pkg.version, arch=ac_pkg.architecture, category=ac_pkg.section, origin=ac_pkg.origins[0].origin)
+
+
+class PIP(CLIMgr):
+
+    CLI = 'pip'
+
+    def list_installed(self):
+        rc, out, err = self.m.run_command([self.cli, 'list', '-l', '--format=json'])
+        if rc != 0:
+            raise Exception("Unable to list packages rc=%s : %s" % (rc, err))
+        return json.loads(out)
 
-    if manager is None or manager == 'auto':
+    def get_package_details(self, package):
+        return package
```

Comment:
This needs to return a dict and needs to return version as well as package name.

Reply:
which it should be for this plugin, which does a json.loads, we can add verification, but i would do that in parent then and use super on all children?

Reply:
That would be a bug in list_installed(), then.  We should be returning the same data types across all the plugins. So list_installed() should be returning a list of package names.

(Probably want to use a @property here as well to cache the json.loads() output)

Reply:
no i mostly designed as 2 functions for one to query the list of packages and the other to properly format, in the case of pip with `--format=json` it all already happens in the first function